Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs

```diff
@@ -480,42 +493,161 @@ private PreTypeResolver(ModuleResolver resolver, PreTypeInferenceModuleState pre
       Constraints.AddConfirmation(check, preType, tok, errorFormatString, onProxyAction);
     }
 
-    void AddComparableConstraint(PreType a, PreType b, IToken tok, string errorFormatString) {
-      Contract.Requires(a != null);
-      Contract.Requires(b != null);
-      Contract.Requires(tok != null);
-      Contract.Requires(errorFormatString != null);
-      Constraints.AddGuardedConstraint(() => ApplyComparableConstraints(a, b, tok, errorFormatString));
+    void AddComparableConstraint(PreType a, PreType b, IToken tok, bool forAsOrIs, string errorFormatString) {
+      // A "comparable types" constraint involves a disjunction. This can get gnarly for inference, so the full disjunction
```

Comment:
This setup sounds gnarly by itself ! Can you elaborate more? Why can't we avoid having duplicate constraints? Maybe the full disjunction check should not emit the duplicate, and then this check can avoid the report false exception? Can you point me to the full disjunction check code?

Reply:
The full disjunction check is in method `CheckComparableTypes` in `PreTypeResolve.cs`.

Reply:
Here's why it's difficult to do the check as part of the constraint machinery:

Suppose the program contains an expression `a == b`, where the types of `a` and `b` are--so far--known to be `R<?0, ?1>` and `R<?2, ?3>`. We would then start off with the comparable-types constraint

    R<?0, ?1>  ~~  R<?2, ?3>

Since the head symbols are the same (`R`), we might hope to turn this constraint into type constraints in terms of the type arguments, like:

    ?0 ~~ ?2    and    ?1 ~~ ?3

But this is not right. For example, if the two type parameters of `R` are co-variant, then these constraints would allow

    ?0 :> ?2    and    ?3 :> ?1

In other words, by turning the original `~~` constraint into `~~` constraints of the type arguments, we lose the correlation that these two are supposed to have.

Reply:
Instead of trying to get the constraint machinery to handle comparable-type constraints completely, it only lays down an approximate constraint (see method `ApproximateComparableConstraints`). In terms of the example above, the approximate constraint only break the `~~` apart if it knows which direction the subtyping is going to go. For example, if it becomes clear that `?0` is a strict subtype of `?2`, then we can, without making any guesses or hasty decisions, lay down the corresponding constraint that `?1` be a subtype of `?3`.

Reply:
When `SolveAllTypeConstraints` runs, it wants to make sure that all constraints have been satisfied. So, if the approximate constraints can't figure everything out, an error would be reported. But `SolveAllTypeConstraints` also calls `ConfirmTypeConstraints()` to confirm things like the full disjunction. We would then end up with two error messages for the same thing. By setting `reportErrors` to `false` for the approximate constraint, we avoid the duplicate error.

Reply:
I understood that without `reportErrors`, the current code would emit a duplicate error, but what I did not understand is why the first error can not prevent the second error from being emitted. Maybe the call `Constraints.AddConfirmation(tok, () => CheckComparableTypes(a, b, forAsOrIs), () => string.Format(errorFormatString, a, b));` can be move so it happens inside `ApproximateComparableConstraints` but only if not error was emitted yet.